site/source/docs/porting/asyncify.rst

```diff
@@ -220,7 +220,10 @@ and want to ``await`` a dynamically retrieved ``Promise``, you can call an
 .. code-block:: cpp
 
     val my_object = /* ... */;
-    val result = my_object.call("someAsyncMethod").await();
+    val result = my_object.call<val>("someAsyncMethod").await();
+
+Note that if this code is invoked from JavaScript then it will be implicitly
+treated as an async function, returning a ``Promise`` for the return value.
```

Comment:
Interesting... I guess this has to work that way, but I'm a little unsure how useful it is. That Promise, what happens if you do a `then()` on it? Does that happen before or after the code resumes? (That is, the code will resume regardless of whether the caller even notices the Promise, so I'm unclear how it interacts.)

A small, complete example of that might be nice to add, if it's not hard (and if the Promise is useful).

Reply:
I thought it had to be this way given in javascript a function with an await must be async and returns a promise, allowing to call to return immediately, was glad to see it did.
To my understanding if you do then() or await on the promise it will run after the cpp .await() had return.
I'm not sure when the decay into promise happens, I think it happens when switching from cpp to javascript, since an intermediate c function behaved just fine and also returned a promise, I'll work my test code for an example.

Reply:
Here is a somewhat more complete example which show that a promise is returned to javascript as soon as the `.await()` is called in cpp, and the promise need to be awaited or `.then()` (perhaps it could be called the law of conservation of async).

```cpp
#include <emscripten/val.h>

using namespace emscripten;

static int fetchData(val obj) {
    printf("fetchData before await\n");
    val result = obj.call<val>("asyncFetch").await();
    printf("fetchData after await\n");
    return result.as<int>();
}

static int getAnswer(val obj) {
    printf("getAnswer before fetch\n");
    int result = fetchData(obj);
    printf("getAnswer after fetch\n");
    return result * 2;
}

EMSCRIPTEN_BINDINGS(answerMachine) {
  emscripten::function("getAnswer", &getAnswer);
}
```
And the following javascript in console:
```javascript
function run() {
    let answer = Module.getAnswer({'asyncFetch' :  async () => 21 });
    console.log(answer instanceof Promise);
    answer.then(console.log);
}
run();
```
yields:
```
getAnswer before fetch
fetchData before await
true
fetchData after await
getAnswer after fetch
42

Reply:
Thanks! Yes, maybe you're right. I am surprised though that not doing an await will stop further execution, but that might be my confusion. @RReverser is that expected?

Reply:
Yes, it's expected. #14664 where the async exports were implemented has some more info on motivation (TL;DR this is done for backward compatibility with functions that don't use Asyncify either directly or indirectly deeper in the call stack).

I agree it's good to clarify this in the docs - thanks for doing it. That said, it's important to note it's not specifically related to `.await()`, but to any Asyncify API, including JS functions implemented via `Asyncify.handle{Sleep, Async}`, standard functions like `emscripten_sleep` and so on. Moreover, such APIs don't have to be at the top level of Embind-exposed function either.

The rule is, if a function has to pause somewhere inside for an Asyncify call, we don't have any useful value to return synchronously yet, so instead we return a Promise that user can await.

Reply:
I'd probably add some note on this to Embind docs instead, as this behaviour is specific to Embind exports, not to any other Asyncify usage.

Reply:
Thanks for clarifying, I've also noted that it depends on the run time, if the .await() isn't invoked then the return value will be the actual one. That said, I'm not sure I understand the last comment in view of the first one, where would this note belong?

Reply:
> Thanks for clarifying, I've also noted that it depends on the run time, if the .await() isn't invoked then the return value will be the actual one.

Yes, this is because Embind doesn't know in advance which functions will end up being async.

In regular JavaScript, to use `await`, you would need to mark every function in the callstack as `async` manually so that the engine knows what to expect and prepares for async behaviour.

Instead, Emscripten's Asyncify specifically tries to make such usage transparent to the C/C++ code, so no changes are necessary and you can implement synchronous C/C++ APIs using async runtime behaviour.

Both of those approaches work great for their respective environments, but lead to pretty much inevitable confusion at integration points like Embind where JS and C++ call into each other.

Reply:
> That said, I'm not sure I understand the last comment in view of the first one, where would this note belong?

There is no definitive place, since this is an intersection of 2 features, but I'm tempted to say it's better to add a section "Integration with Asyncify" in [embind.rst](https://github.com/emscripten-core/emscripten/blob/main/site/source/docs/porting/connecting_cpp_and_javascript/embind.rst) instead, and mention both `.await()` and the export wrapping behaviour over there.

Reply:
Wait, no, I completely missed there is already a section for that in asyncify.rst and that's where you're adding it.

In that case, that's fine too, leave it here but please expand on the behaviour, explaining that it's not limited to `.await()` but rather than any exported function will return a Promise if it invokes any Asyncify API.

Reply:
Ok thanks I'll work it out, and also add the above code as a detailed example.

Reply:
I've updated the doc changes as suggested, please take a look.